### ISSUE TITLE
NHibernate saga throwing error Index IDX_1F02016C already exists

### ISSUE THREAD

**Issue by singhvivek2503** [Particular/NServiceBus.NHibernate#701]
I am using NHibernate persistance for Saga. Its throwing below error:
```
NHibernate.MappingException: Index IDX_1F02016C already exists!
   at NHibernate.Mapping.Table.AddIndex(Index index)
   at NServiceBus.SagaPersisters.NHibernate.AutoPersistence.SagaModelMapper.AddMappings(Configuration configuration, SagaMetadataCollection allSagaMetadata, IEnumerable`1 types, Func`2 tableNamingConvention) in /_/src/NServiceBus.NHibernate/SagaPersisters/SagaModelMapper.cs:line 68
   at NServiceBus.Features.NHibernateSagaStorage.ApplyMappings(ReadOnlySettings settings, Configuration configuration) in /_/src/NServiceBus.NHibernate/SagaPersisters/NHibernateSagaStorage.cs:line 48
   at NServiceBus.Features.NHibernateSagaStorage.<>c__DisplayClass1_0.<Setup>b__0(Configuration configuration) in /_/src/NServiceBus.NHibernate/SagaPersisters/NHibernateSagaStorage.cs:line 30
   at NServiceBus.Features.SharedMappings.<>c__DisplayClass1_0.<ApplyTo>b__0(Action`1 m) in /_/src/NServiceBus.NHibernate/SynchronizedStorage/SharedMappings.cs:line 15
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at NServiceBus.Features.SharedMappings.ApplyTo(Configuration configuration) in /_/src/NServiceBus.NHibernate/SynchronizedStorage/SharedMappings.cs:line 15
   at NServiceBus.Features.NHibernateStorageSession.Setup(FeatureConfigurationContext context) in /_/src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateStorageSession.cs:line 122
   at NServiceBus.Features.FeatureActivator.FeatureInfo.InitializeFrom(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 225
   at NServiceBus.Features.FeatureActivator.ActivateFeature(FeatureInfo featureInfo, List`1 featuresToActivate, FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 188
   at NServiceBus.Features.FeatureActivator.SetupFeatures(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 54
   at NServiceBus.FeatureComponent.Initalize(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureComponent.cs:line 29
   at NServiceBus.EndpointCreator.Initialize() in /_/src/NServiceBus.Core/EndpointCreator.cs:line 59
   at NServiceBus.EndpointCreator.Create(SettingsHolder settings, Configuration hostingConfiguration) in /_/src/NServiceBus.Core/EndpointCreator.cs:line 25
   at NServiceBus.HostCreator.CreateWithExternallyManagedContainer(EndpointConfiguration endpointConfiguration, IConfigureComponents externalContainer) in /_/src/NServiceBus.Core/Hosting/HostCreator.cs:line 34
   at NServiceBus.EndpointWithExternallyManagedContainer.Create(EndpointConfiguration configuration, IConfigureComponents configureComponents) in /_/src/NServiceBus.Core/EndpointWithExternallyManagedContainer.cs:line 20
   at NServiceBus.HostBuilderExtensions.<>c__DisplayClass0_0.<UseNServiceBus>b__0(HostBuilderContext ctx, IServiceCollection serviceCollection) in /_/src/NServiceBus.Extensions.Hosting/HostBuilderExtensions.cs:line 34
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
```

**Comment by ramonsmits** [comment]
@singhvivek2503 Based on the stacktrace it seems to be related to the schema of a saga. Were any changes made to it? NHibernate cannot deal well with modifying an existing schema but only with schema additions. This is not something that NServiceBus can influence.

What would be easiest would be to drop that index and then have it recreated. The error should not persist if NHibernate detects the schema is identical.

Can you share version information? What version of the NHibernate package are you using?

**Comment by singhvivek2503** [comment]
@ramonsmits I am using package NServiceBus 7.5.0 and NServiceBus.Nhibernate 8.5.0

**Comment by ramonsmits** [comment]
@singhvivek2503 

1. Were any changes made to the saga data?
2. Could you drop the index to check if it gets recreated and no longer throw this error?

**Comment by singhvivek2503** [comment]
Hi @ramonsmits,
My schema list looks like below:

```
"SchemaList": {
    "error": "dbo",
    "audit": "dbo",
    "Particular.ServiceControl": "dbo",
    "Particular.Monitoring": "dbo"
  },
```
I am not sure about dropping the index as I cannot find Index with name "IDX_1F02016C" in DB.

**Comment by ramonsmits** [comment]
@singhvivek2503 With schema I meant the full table and index structure (schema) for the sagas used in this endpoint.

1. Were any changes made to the saga data? You have not answered this question yet..
2. The index does not exist? Could you reach out to us via https://particular.net/support with your corporate address to maybe arrange a call as currently I don't have enough details to fully understand your issue and help to resolve it.

**Comment by singhvivek2503** [comment]
@ramonsmits Found the issue. I was using common models between multiple sagas.
While going through the NServicebus documentation found this [Long-running means stateful
](https://docs.particular.net/nservicebus/sagas/#long-running-means-stateful)

> Saga data types should not be shared across different sagas as this can result in persisters physically sharing the same storage structure.

After I separated out models for each saga error was gone. :)

**Comment by ramonsmits** [comment]
@singhvivek2503 ok, so you have several saga types that were partially overlapping.

Interesting that NHibernate throws that error for that state as from an NHibernate perspective that would be allowed but it is a recommendation not to do this as that created coupling between sagas while these should be isolated.
 
Was it something similar like the following snippet?

```c#

public class SagaA : ContainSagaData
{
    public Shared Data { get; set; }
}

public class SagaB : ContainSagaData
{
    public Shared Data { get; set; }
}

public class Shared
{
   public Guid Id { get; set; }
   public string C { get; set; }
   public string D { get; set; }
}
```

**Comment by singhvivek2503** [comment]
Yes, that's how it was.

**Comment by ramonsmits** [comment]
@singhvivek2503 thanks, closing this as your issue has been resolved. Would you have any feedback for us to update our guidance to make this more visible?

**Comment by singhvivek2503** [comment]
Yes please. It was difficult for me to figure this out so it would be great if its more visible in docs. Thanks for your help.

**Comment by cquirosj** [comment]
This improvement has been released as part of version [9.0.1](https://github.com/Particular/NServiceBus.NHibernate/releases/tag/9.0.1)
